app/demo-geant-integration/HepMC3Reader.hh

```diff
@@ -0,0 +1,72 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HepMC3Reader.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <G4Event.hh>
+#include <G4VPrimaryGenerator.hh>
+#include <HepMC3/GenEvent.h>
+#include <HepMC3/GenParticle_fwd.h>
+#include <HepMC3/GenVertex_fwd.h>
+
+// Forward declarations
+namespace HepMC3
+{
+class Reader;
+} // namespace HepMC3
+
+namespace demo_geant
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Minimal HepMC3 reader class. This should be shared among threads
+ */
+class HepMC3Reader final : public G4VPrimaryGenerator
+{
+  public:
+    struct Primary
+    {
+        int           pdg;
+        double        energy;
+        G4ThreeVector vertex;
+        G4ThreeVector momentum;
+    };
+
+    //!@{
+    //! \name Type aliases
+    using Primaries = std::vector<Primary>;
+    //!@}
```

Comment:
These aren't part of the interface so should be private... but I think you can remove them entirely by moving the reading inline with the generating.